### ISSUE TITLE
Abstract repository class and type specification

### ISSUE THREAD

**Issue by gianiaz** [phpstan/phpstan-doctrine#110]
Hi, don't know if this is a bug or an intended behaviour.

We use phpstan (v.0.12.8) and phpstan/phpstan-doctrine(0.12.9), and in our project there's an abstract class (BaseRepository) with some helper methods that extends EntityRepository.

Every repo in our project extends BaseRepository, and for this repositories everything works fine thanks to the extension.

I have this error on BaseRepository class:

```
------ --------------------------------------------------------------------- 
  Line   src/Core/Repository/BaseRepository.php                               
 ------ --------------------------------------------------------------------- 
  8      Class Core\Repository\BaseRepository extends generic   
         class Doctrine\ORM\EntityRepository but does not specify its types:  
         TEntityClass                                                         
         💡 You can turn this off by setting                                   
         checkGenericClassInNonGenericObjectType: false in your               
         phpstan.neon.                                                        
 ------ ------------------------------------------------------------------
```

Imho this error should not trigger with abstract classes.

**Comment by ondrejmirtes** [comment]
It should. Please see the section "Class-level generics" here: https://medium.com/@ondrejmirtes/generics-in-php-using-phpdocs-14e7301953

> When implementing a generic interface or extending a generic class, you have two options:
> Preserve the genericness of the parent, the child class will also be generic
> Specify the type variable of the interface/parent class. The child class will not be generic.

**Comment by gianiaz** [comment]
Sorry to bother again on this thing, but I need an advice.

As I said I have:

ConcreteRepository extends BaseRepository 

Abstract BaseRepository extends EntityRepository.

On BaseRepository I've put this doc:

```php
/**
 * @template T
 * @extends EntityRepository<T>
 */
abstract class BaseRepository extends EntityRepository
```

Now I have a lot of errors on concreteRepository that not specify its type.
This is the job done through the phpstan-doctrine extension, I am right? 
If I omit the above doc I have a single error on baseRepository. 

What should be the most elegant way to document this behavior without specifying the type on every repository? 

Thank you

**Comment by ondrejmirtes** [comment]
If your repositories are 1:1 mapped to your entities, you should document it above each repository class. The only way around it is to disable the check, but you're losing strong typing and PHPStan cannot do the best job it's capable of.

**Comment by gianiaz** [comment]
Thank you very much @ondrejmirtes

**Comment by TomasVotruba** [comment]
I'm trying to define `TEntityClass` on constructor injection approach, but don't know how.

```php
<?php

final class UserRepository
{
    /**
     * @template TEntityClass of User
     * @extends \Doctrine\Persistence\ObjectRepository<TEntityClass>
     */
    private ObjectRepository $repository;

    public function __construct(EntityManagerInterface $entityManager)
    {
        $this->repository = $entityManager->getRepository(User::class);
    }

    public function findByLogin(string $login): ?User
    {
        return $this->repository->findOneBy([
            'login' => $login,
        ]);
    }
}
```

Where should I put `TEntityClass` it?

**Comment by ondrejmirtes** [comment]
You just need `@var \Doctrine\Persistence\ObjectRepository<User>` above the property, nothing else.

**Comment by TomasVotruba** [comment]
Works, thanks :+1:
